gc/default/default.c

```diff
@@ -5471,14 +5487,21 @@ gc_compact_move(rb_objspace_t *objspace, rb_heap_t *heap, VALUE src)
         }
     }
 
+    VALUE dest = rb_gc_impl_location(objspace, src);
+
     if (orig_shape != 0) {
         if (new_shape != 0) {
-            VALUE dest = rb_gc_impl_location(objspace, src);
             rb_gc_set_shape(dest, new_shape);
         }
         RMOVED(src)->original_shape_id = orig_shape;
     }
-
+    if (old_address) {
+        DURING_GC_COULD_MALLOC_REGION_START();
+        {
+            rb_obj_field_set(dest, old_address_shape, old_address);
```

Comment:
NB: With the introduction of namespaces, now classes potentially have one set of fields per namespaces, so storing the address in fields is no longer possible.

That being said, from my understanding this only apply to "core" classes, which I suspect are all pinned, so we could probably still get away with it, but we'd have to be careful.